### PR TITLE
feat(rest): allow AJV error messages customization

### PR DESCRIPTION
**Pull request by AaronJan**
I'm currently doing some localization works, it should be easy by just using [`ajv-i18n`](https://github.com/epoberezkin/ajv-i18n):

```typescript
const localize_zh = require('ajv-i18n/localize/zh');

// option `i18n` is required for this package to work
var ajv = Ajv({ allErrors: true });
var validate = ajv.compile(schema);
var valid = validate(data);

if (!valid) {
    localize_zh(validate.errors);
    console.log(ajv.errorsText(validate.errors, { separator: '\n' }));
}
```

But I could find a way to do that.

After some searching & reading, I found #2146, I really need this feature, so I'm submitting a minimal change.

Advice needed (we can discus in #2146 ), thank you!

> See also #2146 

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/rest/src/types.ts [hunk 1]
```diff
@@ -107,6 +107,10 @@ export interface RequestBodyValidationOptions extends ajv.Options {
    * - `string[]`: Add an array of keywords from `ajv-keywords`
    */
   ajvKeywords?: true | string[];
+  /**
+   * Enable error message customizing.
+   */
+  ajvTranslate?: (errors: ajv.ErrorObject[]) => ajv.ErrorObject[];
```
Comment (raymondfeng) [inline packages/rest/src/types.ts]: Should we allow asynchronous translation here?

Comment (AaronJan) [reply packages/rest/src/types.ts]: @raymondfeng Yes, it would be more unified, but I think I have to touch a couple more files to do that.

Is that OK?

Comment (bajtos) [reply packages/rest/src/types.ts]: So far, we support synchronous validation only, therefore `validateValueAgainstSchema` is sync too. I feel introducing async translation would require too many changes, I am not convinced the (speculative) benefits are worth the cost.

Let's start with sync translations for the first version, we can look into async translations later, if there is need.

Comment (bajtos) [inline packages/rest/src/types.ts]: I like the idea of allowing app developers to transform the errors reported by AJV. I am little bit reluctant to limit this new config option to translations only, can find a more generic name please? For example: `ajvErrorTransformer`.

File: packages/rest/src/__tests__/acceptance/validation/validation.acceptance.ts [hunk 2]
```diff
@@ -246,6 +246,64 @@ describe('Validation at REST level', () => {
...
+            return errors.map(e => ({
+              ...e,
+              message: `(translated) ${e.message}`,
+            }));
+          },
+        }),
+      );
+      after(() => app.stop());
+
+      it('rejects requests with translated error message', async () => {
```
Comment (bajtos) [inline packages/rest/src/__tests__/acceptance/validation/validation.acceptance.ts]: I read this test name as "the framework reject requests when the client provided a translated error message". Let's fine a better one please. For examples:

```suggestion
      it('transforms error messages provided by AJV', async () => {
```

File: packages/rest/src/validation/request-body.validator.ts [hunk 3]
```diff
@@ -137,7 +137,11 @@ function validateValueAgainstSchema(
...
 
-  const validationErrors = validate.errors;
+  let validationErrors = validate.errors as AJV.ErrorObject[];
+
+  if (options.ajvTranslate) {
+    validationErrors = options.ajvTranslate(validationErrors);
+  }
 
   /* istanbul ignore if */
   if (debug.enabled) {
```
Comment (bajtos) [inline packages/rest/src/validation/request-body.validator.ts]: The next line prints a debug log with `util.inspect(validationErrors)`.

In your current proposal, these debug logs will contain translated error message. This is nice when there is a single language involved only and the developer of the application understand the language used for messages returned to clients.

It does not work well for applications supporting multiple languages - for example, if I was developing an app available in English, Czech and Chinese, then I won't be able to understand debug logs translated to Chinese. Personally, I'd prefer to see all debug logs in English using the original message reported by AJV, before any translation or other transformations is applied.

Please run the transformation after the debug log is printed.

If you feel it's useful to see translated messages in the debug log too, then add a new debug log that's printed after the AJV errors were modified.

```ts
if (options.ajvTranslate) {
  validationErrors = options.ajvTranslate(validationErrors);
  /* istanbul ignore if */
  if (debug.enabled) {
    debug(...);
  }
}
```

File: packages/rest/src/validation/request-body.validator.ts [hunk 4]
```diff
@@ -137,7 +137,11 @@ function validateValueAgainstSchema(
     return;
   }
 
-  const validationErrors = validate.errors;
+  let validationErrors = validate.errors as AJV.ErrorObject[];
+
+  if (options.ajvTranslate) {
```
Comment (bajtos) [inline packages/rest/src/validation/request-body.validator.ts]: Let's be little bit more defensive and check that `ajvTranslate` is actually a function. We don't want to see `TypeError: options.ajvTranslate is not a function`.

```suggestion
  if (typeof options.ajvTranslate === 'function') {
```

File: packages/rest/src/types.ts (hunk 1 above)
Comment (AaronJan) [reply packages/rest/src/types.ts]: I was uncertain about using verb or noun, I think `ajvErrorTransformer ` is good. 👍

Comment (AaronJan) [reply packages/rest/src/types.ts]: I was worried about the probably low ROI of this async change, but I don't like to support both sync & async transformation in the future either.

But I think minimizing changes is better for now.

**Comment by AaronJan** [comment]
I made a new commit according @bajtos 's suggestions.

Let me know if there is anything can be improved or rebasing. 😄

File: packages/rest/src/__tests__/acceptance/validation/validation.acceptance.ts [hunk 5]
```diff
@@ -246,6 +246,64 @@ describe('Validation at REST level', () => {
     });
   });
 
+  context(
+    'with request body validation options - {ajvTranslate: [Function]}',
```
Comment (raymondfeng) [inline packages/rest/src/__tests__/acceptance/validation/validation.acceptance.ts]: `ajvTranslate` => `ajvErrorTransformer`

**Comment by bajtos** [comment]
Could you @AaronJan rebase your patch on top of the latest master and squash all commits into a single one please? See [Rebasing](https://loopback.io/doc/en/lb4/submitting_a_pr.html#6-rebasing) and [Final rebase and squashing of commits](https://loopback.io/doc/en/lb4/submitting_a_pr.html#10-final-rebase-and-squashing-of-commits) in our guide. See also our [Commit message guidelines](https://loopback.io/doc/en/lb4/code-contrib-lb4.html#commit-message-guidelines).

**Comment by AaronJan** [comment]
Done. @bajtos

**Pull request merged**